clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java

```
@@ -0,0 +1,119 @@
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.config.ConfigException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.kafka.common.requests.InitPIDResponse.INVALID_PID;
+
+/**
+ * A class which maintains state for transactions. Also keeps the state necessary to ensure idempotent production.
+ */
+public class TransactionState {
+    String transactionalId;
+    private volatile long pid;
+    private short epoch;
+    private boolean isStateValid;
+    private final boolean idempotenceEnabled;
+    Map<TopicPartition, Integer> sequenceNumbers;
+
+
+    public TransactionState(String transactionalId, boolean idempotenceEnabled) {
+        if (transactionalId != null && 0 < transactionalId.length() && !idempotenceEnabled) {
+            throw new ConfigException("Need to set " + ProducerConfig.IDEMPOTENCE_ENABLED_CONFIG
+                    + " to true in order to use transactions.");
+        }
+        this.transactionalId = transactionalId;
+        pid = INVALID_PID;
+        epoch = 0;
+        sequenceNumbers = new HashMap<>();
+        this.isStateValid = true;
+        this.idempotenceEnabled = idempotenceEnabled;
+    }
+
+    public TransactionState() {
+        this(false);
+    }
+
+    public TransactionState(boolean idempotenceEnabled) {
+        this(null, idempotenceEnabled);
+    }
+
+    public boolean pidIsSet() {
+        return pid() != INVALID_PID;
+    }
+
+    public Long pid() {
+        return pid;
+    }
+
+    public short epoch() {
+        return epoch;
+    }
+
+    public boolean isIdempotenceEnabled() {
+        return idempotenceEnabled;
+    }
+
+    public synchronized void setPid(long pid) {
```

We don't need this to be synchronized, since `setPid` is only called by the background thread. Also, we should raise an exception if `setPid` is called when `pid` is already set (ie. it is anything but `INVALID_PID`).

However, the transaction state is accessed by multiple threads. In particular, the `sequenceNumber` and `incrementSequenceNumber` are called in the `RecordAccumulator` and `RecordBatch`, which execute in the context of application threads. I think those methods need to be synchronized instead.